From 9d6bfb4a5e9b44c080ddf3bad4c364ffb0e9d84a Mon Sep 17 00:00:00 2001
From: Dengke Du <dengke.du@windriver.com>
Date: Wed, 8 May 2019 10:20:47 +0800
Subject: [PATCH] Add 'install-ptest' rule

Change TESTS_ENVIRONMENT to allow running outside build dir.

Upstream-status: Pending
Signed-off-by: Mihaela Sendrea <mihaela.sendrea@enea.com>
[KK: Update context for 1.3.5.]
Signed-off-by: Kai Kang <kai.kang@windriver.com>
[MA: Allow separate source and build dirs]
Signed-off-by: Mark Asselstine <mark.asselstine@windriver.com>
[ZH: add missing test_helper files]
Signed-off-by: He Zhe <zhe.he@windriver.com>
[MA: Update context for v4.3.0]
Signed-off-by: Mark Asselstine <mark.asselstine@windriver.com>
[DDU: Update context for v5.3.0]
Signed-off-by: Dengke Du <dengke.du@windriver.com>

---
 tests/Makefile.am | 68 +++++++++++++++++++++++++++++++++++++++++------
 1 file changed, 60 insertions(+), 8 deletions(-)

diff --git a/tests/Makefile.am b/tests/Makefile.am
index 83326db..7a0aee3 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -28,11 +28,13 @@ AM_CPPFLAGS = \
 
 WARN_CFLAGS += $(RELAXED_FRAME_LIMIT_CFLAGS)
 
+PTEST_DIR ?= $(libdir)/libvirt/ptest
+
 AM_CFLAGS = \
-	-Dabs_builddir="\"$(abs_builddir)\"" \
-	-Dabs_top_builddir="\"$(abs_top_builddir)\"" \
-	-Dabs_srcdir="\"$(abs_srcdir)\"" \
-	-Dabs_top_srcdir="\"$(abs_top_srcdir)\"" \
+	-Dabs_builddir="\"$(PTEST_DIR)/tests\"" \
+	-Dabs_top_builddir="\"$(PTEST_DIR)\"" \
+	-Dabs_srcdir="\"$(PTEST_DIR)/tests\"" \
+	-Dabs_top_srcdir="\"$(PTEST_DIR)\"" \
 	$(LIBXML_CFLAGS) \
 	$(GLIB_CFLAGS) \
 	$(LIBNL_CFLAGS) \
@@ -472,10 +474,10 @@ TESTS = $(test_programs) \
 
 VIR_TEST_EXPENSIVE ?= $(VIR_TEST_EXPENSIVE_DEFAULT)
 TESTS_ENVIRONMENT = \
-  abs_top_builddir="$(abs_top_builddir)" \
-  abs_top_srcdir="$(abs_top_srcdir)" \
-  abs_builddir="$(abs_builddir)" \
-  abs_srcdir="$(abs_srcdir)" \
+  abs_top_builddir="$(PTEST_DIR)" \
+  abs_top_srcdir="$(PTEST_DIR)" \
+  abs_builddir="$(PTEST_DIR)/tests" \
+  abs_srcdir="$(PTEST_DIR)/tests" \
   LIBVIRT_AUTOSTART=0 \
   LC_ALL=C \
   VIR_TEST_EXPENSIVE=$(VIR_TEST_EXPENSIVE) \
@@ -1549,4 +1551,54 @@ else ! WITH_LINUX
 EXTRA_DIST += virscsitest.c
 endif  ! WITH_LINUX
 
+buildtest-TESTS: $(TESTS) $(test_libraries) $(test_helpers)
+
+PTESTS = $(TESTS) $(test_helpers) test-lib.sh virschematest
+
+install-ptest:
+	list='$(TESTS) $(test_helpers) test-lib.sh virschematest'
+	install -d $(DEST_DIR)/tools
+	@(if [ -d ../tools/.libs ] ; then cd ../tools/.libs; fi; \
+	install * $(DEST_DIR)/tools)
+	install -d $(DEST_DIR)/src/network
+	cp $(top_srcdir)/src/network/*.xml $(DEST_DIR)/src/network
+	install -d $(DEST_DIR)/src/cpu_map
+	cp $(top_srcdir)/src/cpu_map/*.xml $(DEST_DIR)/src/cpu_map
+	install ../src/libvirt_iohelper $(DEST_DIR)/src
+	install -D ../src/libvirtd $(DEST_DIR)/src/libvirtd
+	install -d $(DEST_DIR)/src/remote
+	install -D $(top_srcdir)/../build/src/remote/libvirtd.conf $(DEST_DIR)/src/remote/libvirtd.conf
+	install -d $(DEST_DIR)/src/remote/.libs
+	@(if [ -d ../src/remote/.libs ] ; then cd ../src/remote/.libs; fi; \
+	install * $(DEST_DIR)/src/remote/.libs)
+	install -d $(DEST_DIR)/src/.libs
+	@(if [ -d ../src/.libs ] ; then cd ../src/.libs; fi; \
+	install * $(DEST_DIR)/src/.libs)
+	install -d $(DEST_DIR)/docs/schemas
+	cp $(top_srcdir)/docs/schemas/*.rng $(DEST_DIR)/docs/schemas
+	cp -r $(top_srcdir)/build-aux $(DEST_DIR)
+	install -d $(DEST_DIR)/examples/xml
+	cp -r $(top_srcdir)/examples/xml/test $(DEST_DIR)/examples/xml
+	install -d $(DEST_DIR)/tests/.libs
+	find . -type d -name "*xml2xml*" -exec cp -r {} $(DEST_DIR)/tests \;
+	find . -type d -name "*data" -exec cp -r {} $(DEST_DIR)/tests \;
+	@(for file in $(PTESTS); do \
+		if [ -f .libs/$$file ]; then \
+			install .libs/$$file $(DEST_DIR)/tests; \
+		elif [ -f $(srcdir)/$$file ]; then \
+			install $(srcdir)/$$file $(DEST_DIR)/tests; \
+		else \
+			install $(builddir)/$$file $(DEST_DIR)/tests; \
+		fi; \
+	done;)
+	@(if [ -d .libs ]; then install .libs/*.so $(DEST_DIR)/tests/.libs; fi;)
+	cp ../config.h $(DEST_DIR)
+	cp Makefile $(DEST_DIR)/tests
+	sed -i -e 's/^Makefile:/_Makefile:/' $(DEST_DIR)/tests/Makefile
+	cp ../Makefile $(DEST_DIR)
+	sed -i -e 's|^Makefile:|_Makefile:|' $(DEST_DIR)/Makefile
+	sed -i -e 's|$(BUILD_DIR)|$(PTEST_DIR)|g' $(DEST_DIR)/tests/Makefile
+	sed -i -e 's|$(BUILD_DIR)|$(PTEST_DIR)|g' $(DEST_DIR)/Makefile
+	sed -i -e 's|^\(.*\.log:\) \(.*EXEEXT.*\)|\1|g' $(DEST_DIR)/tests/Makefile
+
 CLEANFILES = *.cov *.gcov .libs/*.gcda .libs/*.gcno *.gcno *.gcda
